cpp/src/arrow/dataset/expression.cc

```diff
@@ -0,0 +1,1177 @@
+    if (!arg.IsScalarExpression()) return false;
+  }
+
+  if (call->function) {
+    return call->function->kind() == compute::Function::SCALAR;
+  }
+
+  // this expression is not bound; make a best guess based on
+  // the default function registry
+  if (auto function = compute::GetFunctionRegistry()
+                          ->GetFunction(call->function_name)
+                          .ValueOr(nullptr)) {
+    return function->kind() == compute::Function::SCALAR;
+  }
+
+  // unknown function or other error; conservatively return false
+  return false;
+}
+
+bool Expression::IsNullLiteral() const {
+  if (auto lit = literal()) {
+    if (lit->null_count() == lit->length()) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool Expression::IsSatisfiable() const {
+  if (descr().type && descr().type->id() == Type::NA) {
+    return false;
+  }
+
+  if (auto lit = literal()) {
+    if (lit->null_count() == lit->length()) {
+      return false;
+    }
+
+    if (lit->is_scalar() && lit->type()->id() == Type::BOOL) {
+      return lit->scalar_as<BooleanScalar>().value;
+    }
+  }
+
+  if (auto ref = field_ref()) {
+    return true;
+  }
+
+  return true;
+}
+
+inline bool KernelStateIsImmutable(const std::string& function) {
+  // XXX maybe just add Kernel::state_is_immutable or so?
+
+  // known functions with non-null but nevertheless immutable KernelState
+  static std::unordered_set<std::string> names = {
+      "is_in", "index_in", "cast", "struct", "strptime",
+  };
+
+  return names.find(function) != names.end();
+}
+
+Result<std::unique_ptr<compute::KernelState>> InitKernelState(
+    const Expression::Call& call, compute::ExecContext* exec_context) {
+  if (!call.kernel->init) return nullptr;
+
+  compute::KernelContext kernel_context(exec_context);
+  auto kernel_state = call.kernel->init(
+      &kernel_context, {call.kernel, GetDescriptors(call.arguments), call.options.get()});
+
+  RETURN_NOT_OK(kernel_context.status());
+  return std::move(kernel_state);
+}
+
+Status MaybeInsertCast(std::shared_ptr<DataType> to_type, Expression* expr) {
+  if (expr->descr().type->Equals(to_type)) {
+    return Status::OK();
+  }
+
+  if (auto lit = expr->literal()) {
+    ARROW_ASSIGN_OR_RAISE(Datum new_lit, compute::Cast(*lit, to_type));
+    *expr = literal(std::move(new_lit));
+    return Status::OK();
+  }
+
+  // FIXME the resulting cast Call must be bound but this is a hack
+  auto with_cast = call("cast", {literal(MakeNullScalar(expr->descr().type))},
+                        compute::CastOptions::Safe(to_type));
+
+  static ValueDescr ignored_descr;
+  ARROW_ASSIGN_OR_RAISE(with_cast, with_cast.Bind(ignored_descr));
+
+  auto call_with_cast = *CallNotNull(with_cast);
+  call_with_cast.arguments[0] = std::move(*expr);
+  call_with_cast.descr = ValueDescr{std::move(to_type), expr->descr().shape};
+
+  *expr = Expression(std::move(call_with_cast));
+  return Status::OK();
+}
+
+Status InsertImplicitCasts(Expression::Call* call) {
```

Comment:
The intent is that basic functionality (Binding, Execution) of each compute Function is supported without boilerplate. Special cases and added functionality can then be added as needed for other compute functions, like dictionary decoding for `is_in` or informative `Expression::ToString`. For any function not mentioned in the implicit cast step, binding will simply error out when types are not exactly aligned.

You're correct that I've essentially extended the interface of FunctionOptions, but I want to be careful about moving these features upstream since some of them could/should be implemented elsewhere. For example, ideally the `is_in` function iteslf should support dictionary directly and should handle casting of an improperly typed value_set in DispatchBest. After that change `FunctionOptions::CastFor` would be obviated

Reply:
Sounds reasonable for now.